src/io/binary_parser.cpp

```diff
@@ -161,21 +162,32 @@ long long read_factors(string filename, dd::FactorGraph &fg) {
       }
 
       // add variables to factors
-      fg.factors[fg.c_nfactor].tmp_variables.push_back(dd::VariableInFactor(
+      fg.factors[fg.c_nfactor].add_variable_in_factor(dd::VariableInFactor(
           variable_id, position, ispositive, equal_predicate));
-      fg.variables[variable_id].tmp_factor_ids.push_back(fg.c_nfactor);
+      fg.variables[variable_id].add_factor_id(fg.c_nfactor);
     }
 
     switch (type) {
       case (dd::FUNC_SPARSE_MULTINOMIAL): {
         long n_weights = 0;
         file.read((char *)&n_weights, 8);
         n_weights = be64toh(n_weights);
-        fg.factors[fg.c_nfactor].weight_ids.reserve(n_weights);
-        for (long j = 0; j < n_weights; j++) {
+
+        fg.factors[fg.c_nfactor].weight_ids = new std::unordered_map<long, long>();
+
+        for (long i = 0; i < n_weights; i++) {
+          // calculate radix-based key into weight_ids (see also FactorGraph::get_multinomial_weight_id)
+          long key = 0;
+          for (long j = 0; j < edge_count; j++) {
+            const dd::Variable &var = fg.variables[fg.factors[fg.c_nfactor].tmp_variables->at(j).vid];
+            file.read((char *)&value_id, 4);
+            value_id = be32toh(value_id);
+            key *= var.cardinality;
+            key += var.get_domain_index(value_id);
+          }
```

Comment:
Is there any better way to share this multi-radix-based key calculation code between two?

https://github.com/HazyResearch/sampler/pull/27/files#diff-ee993b6f60b24e71935cf552245e742aR94